File: src/amtsub/sources/certificate_search.py
import requests

class CertificateSearch:
	async def search(self, domain, timeout):
		try:
			url = f'https://crt.sh/?q={domain}&output=json'

			response = requests.get(url, timeout = timeout)

			certificates = response.json()

			subdomains = []

			for certificate in certificates:
				common_name = certificate['common_name']

				subdomains.append(common_name)

				name_value = certificate['name_value']

				lines = name_value.splitlines()

				for line in lines:
					subdomains.append(line)

			return subdomains
		except:
			pass

	def get_name(self):
		return 'Certificate Search'

File: src/amt/sub/sources/certificate_search/certificate_search.go
package certificate_search

import (
	"time"
	"strings"
	"net/http"
)

import (
	jsoniter "github.com/json-iterator/go"
)

type CertificateSearch struct {}

type certificateObj struct {
	NameValue string `json:"name_value"`
	CommonName string `json:"common_name"`
}

func (self CertificateSearch) Search(domain string, timeOut time.Duration) ([]string, error) {
	client := http.Client {
		Timeout: timeOut,
	}

	url := "https://crt.sh/?q=" + domain + "&output=json"

	response, err := client.Get(url)

	if err != nil {
		return []string{}, err
	}

	defer response.Body.Close()

	var json = jsoniter.ConfigCompatibleWithStandardLibrary

	decoder := json.NewDecoder(response.Body)

	var certificates []certificateObj

	err = decoder.Decode(&certificates)

	if err != nil {
		return []string{}, err
	}

	var subdomains []string

	for _, certificate := range certificates {
		commonName := certificate.CommonName

		subdomains = append(subdomains, commonName)

		nameValue := certificate.NameValue

		values := strings.SplitSeq(nameValue, "\n")

		for value := range values {
			subdomains = append(subdomains, value)
		}
	}

	return subdomains, nil
}

func (self CertificateSearch) GetName() string {
	return "Certificate Search"
}